### ISSUE TITLE
do compression for video as well

### ISSUE THREAD

**Issue by grill05** [ayastreb/bandwidth-hero#35]
This extension uses a heroku instance to change jpg/png to .webp. Some sites now have short preview videos which are typically encoded in H.264.
I would like to save bandwidth with videos as well, by converting h.264 to h.265. The conversion will typically reduce video size by 50-65%.
For example, assuming the heroku instance has "ffmpeg" command available, a sample convert would look like
`ffmpeg -i input.mp4 -c:v hevc output.mp4`
and the h.265 video (output.mp4) would be sent to chrome, which can be played at the user's end (using https://chrome.google.com/webstore/detail/h265-hevc-player/dambgipgbnhmnkdolkljibpcbocimnpd?hl=en for example)

**Comment by brianvanburken** [comment]
This sounds like a great idea! 

An approach is to change this extension to also catch videos. First add `media` (https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/ResourceType) to catch video and audio. Then the extension must determine if the URL is a video link. Then it should redirect the URL to the proxy with an added flag to signal the proxy that it is a video. In the proxy (https://github.com/ayastreb/bandwidth-hero-proxy), the video must be downloaded, converted, and send back. I think the proxy must supported streaming so that it won't reach limits on heroku free tier.

It's best to start with figuring out if the proxy can support the downloading, conversion and maybe streaming within a limited heroku instance (free tier).

**Comment by mandaputtra** [comment]
I actually trying to compress youtube video when streaming. Still no luck and the server itselft are too hot (using much ram) to process the compresion. I use [fluent-ffmpeg](https://github.com/fluent-ffmpeg/node-fluent-ffmpeg)

And youtube player would not want to play from another source.

But compress the video and play using another player may success

**Comment by mandaputtra** [comment]
Maybe the extension could be rewritten as a VPN, and then uses [this](https://github.com/barnacs/compy) as a server / create nodejs version of it. I could but for now I don't had any time to understand some Go code since I'm not familiar enough.

**Comment by grill05** [comment]
@mandaputtra I agree that arbitrary video compression would be a problem for a server, especially a heroku free tier instance. I mainly had the case of "preview" videos in mind. Like in youtube, if you hover the mouse over a thumbnail, sometimes a short preview video will be played. The preview itself is only a few seconds long. So compressing it shouldn't put too much load on the server.

**Comment by mandaputtra** [comment]
@grill05 That could be done, will look at it if I had the time. I assume that we must capture `media` on `beforeRequestHandler` too. In my case, the YouTube video player doesn't return media type headers because it's streamed, but maybe the preview are `media` type content.

**Comment by mandaputtra** [comment]
@grill05 I Just look at how the YouTube video streamed, they actually encode all the request so it its not possible even for the youtube preview haha. However for Stock video that request response can be filtered by media it can be done.

**Comment by changhaitravis** [comment]
I have a pull request in bandwidth-hero-proxy for handling and compressing audio and video in <video> and <audio> tags and it works fairly well. However a lot of sites use .m3u for HLS/DASH, and for whatever reason this method doesn't work. I'm seeing some success parsing the .m3u file and then removing specific high-bandwidth or HD stream entries so that it'll select from the remaining low-bandwidth stream even though you are currently on LTE or Cable and it detects you have a ton of bandwidth.
